pkg/compactor/block_upload.go

```diff
@@ -258,55 +268,47 @@ func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Re
 	w.WriteHeader(http.StatusOK)
 }
 
-func decodeMeta(r io.Reader, name string) (metadata.Meta, error) {
-	dec := json.NewDecoder(r)
-	var meta metadata.Meta
-	if err := dec.Decode(&meta); err != nil {
-		return meta, errors.Wrap(err, fmt.Sprintf("failed decoding %s", name))
-	}
-
-	return meta, nil
-}
-
-func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.Request,
-	logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID) error {
-	level.Debug(logger).Log("msg", "received request to complete block upload", "content_length", r.ContentLength)
+func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, logger log.Logger, userBkt objstore.Bucket, blockID ulid.ULID, meta metadata.Meta) {
+	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
 
-	uploadingMetaPath := path.Join(blockID.String(), uploadingMetaFilename)
-	rdr, err := userBkt.Get(ctx, uploadingMetaPath)
-	if err != nil {
-		if userBkt.IsObjNotFoundErr(err) {
-			return httpError{
-				message:    fmt.Sprintf("upload of block %s not started yet", blockID),
-				statusCode: http.StatusNotFound,
-			}
-		}
-		return errors.Wrap(err, fmt.Sprintf("failed to download %s from object storage", uploadingMetaFilename))
+	// create validation file
+	if err := c.uploadValidation(ctx, logger, blockID, userBkt); err != nil {
+		level.Error(logger).Log("msg", "error creating validation file in object store", "blockID", blockID)
+		return
 	}
-	defer func() {
-		_ = rdr.Close()
-	}()
+	// start go routine that updates validation file timestamp once per minute
+	ch := make(chan string)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go c.periodicValidationUpdater(ctx, logger, blockID, userBkt, ch, &wg)
 
-	meta, err := decodeMeta(rdr, uploadingMetaFilename)
-	if err != nil {
-		return err
+	if err := c.validateBlock(ctx, blockID, userBkt, meta); err != nil {
+		err := c.uploadValidationWithError(ctx, logger, blockID, userBkt, err.Error())
+		if err != nil {
+			level.Error(logger).Log("msg", "error updating validation file after failed validation in object store", "blockID", blockID)
+		}
+		return
 	}
 
-	level.Debug(logger).Log("msg", "completing block upload", "files", len(meta.Thanos.Files))
+	// use waitgroup to ensure validation ts update is complete
+	ch <- "Done"
+	wg.Wait()
 
 	// Upload meta file so block is considered complete
 	if err := c.uploadMeta(ctx, logger, meta, blockID, block.MetaFilename, userBkt); err != nil {
-		return err
+		//return err
+		// set error message in validation file
 	}
 
-	if err := userBkt.Delete(ctx, uploadingMetaPath); err != nil {
+	if err := userBkt.Delete(ctx, path.Join(blockID.String(), uploadingMetaFilename)); err != nil {
 		level.Warn(logger).Log("msg", fmt.Sprintf(
 			"failed to delete %s from block in object storage", uploadingMetaFilename), "err", err)
-		return nil
+		//return nil
+		// set error message in validation file
```

Comment:
At this point `meta.json` file for the block exists (thanks to previous `uploadMeta` call), so block upload is complete. If we fail to delete `uploading-meta.json` it is unfortunate, but harmless, and we should NOT set error in validation file.